### ISSUE TITLE
Train ArcFace on my own data from scratch

### ISSUE THREAD

**Issue by beybars1** [deepinsight/insightface#2176]
Hi there,

I find the process of model training from scratch very confusing. Can anybody provide guides/advices/tips and other sort of help on how to train **ArcFace by insightface**, please?

I have already read this [post](https://github.com/deepinsight/insightface/issues/791), but still do not get a clear picture of this process.

Any help will be much appreciated.

**Comment by marlowinnovations** [comment]
Hope this helps you out:

1) Create a folder for your dataset.  Separate your unique classes of images into their own folders inside of the dataset folder
2) Run im2rec.py to build a .lst file of your training data

python3 im2rec.py <list_file_output_prefix> /path/to/training/data/folder --list --recursive
python3 im2rec.py mydata /path/to/training/data/folder --list --recursive (creates mydata.lst)

This will recursively parse the images in the training data folder and build the list file.

3) Build your .rec database using the newly created .lst file

python3 im2rec.py <rec_file_output_prefix> /path/to/output/your/rec/file
python3 im2rec.py mydata /path/to/output/your/rec/file (creates mydata.rec)

**Comment by beybars1** [comment]
> Hope this helps you out:
> 
>     1. Create a folder for your dataset.  Separate your unique classes of images into their own folders inside of the dataset folder
> 
>     2. Run im2rec.py to build a .lst file of your training data
> 
> 
> python3 im2rec.py <list_file_output_prefix> /path/to/training/data/folder --list --recursive python3 im2rec.py mydata /path/to/training/data/folder --list --recursive (creates mydata.lst)
> 
> This will recursively parse the images in the training data folder and build the list file.
> 
>     3. Build your .rec database using the newly created .lst file
> 
> 
> python3 im2rec.py <rec_file_output_prefix> /path/to/output/your/rec/file python3 im2rec.py mydata /path/to/output/your/rec/file (creates mydata.rec)

Thanks for your detailed response. I will try it out and come back with results ASAP.

**Comment by Nomiluks** [comment]
can anyone please guide me to where I can find im2rec.py?

**Comment by beybars1** [comment]
> can anyone please guide me to where I can find im2rec.py?

Here it is, [link](https://github.com/apache/mxnet/blob/master/tools/im2rec.py)

**Comment by youngjae-avikus** [comment]
@beybars1 
@Nomiluks 
@marlowinnovations 
Does it have to be datasets with face detected (maybe body not visible...) and aligned in dst points?

**Comment by marlowinnovations** [comment]
> @beybars1 @Nomiluks @marlowinnovations Does it have to be datasets with face detected (maybe body not visible...) and aligned in dst points?

Crop and align into 112x112 images and then you can use im2rec

**Comment by youngjae-avikus** [comment]
@marlowinnovations 
Thank you for your fast reply...
If I had to ask one more question, I would like to know whether cropping means doing it without considering the aspect-ratio with padding, or whether it is done by maintaining it.

**Comment by marlowinnovations** [comment]
> @marlowinnovations Thank you for your fast reply... If I had to ask one more question, I would like to know whether cropping means doing it without considering the aspect-ratio with padding, or whether it is done by maintaining it.

Yah you'll want to keep the proper aspect ratio so the images don't get distorted.  Whatever face alignment code you plan to use during inference, use the same one to crop and align your images for the dataset so its consistent.

**Comment by karmelyoei** [comment]
@beybars1  Hey, did you succeed in training arcface on your dataset? If so, is there any kind of full explanation or guidelines to follow??

**Comment by beybars1** [comment]
> @beybars1 Hey, did you succeed in training arcface on your dataset? If so, is there any kind of full explanation or guidelines to follow??

Hi!

Yes, I managed to train it, pretty successfully. Let's make it clear once again, for those who is struggling right now and those who probably will...

The general pipeline of any modern SOTA face recognition algorithms can be grouped as following:

1. Face detection phase (You never know when exactly the person's will appear on the screen/camera - thus you need to detect a face.)
2. Face alignment phase (The detected face is not that useful on its own, since different angles/tilts/yaws of detected faces make a significant distribution shift, we do not want it). In this phase, the detected face has to be aligned via particular warp transformation, which exploits facial landmarks (In case of ArcFace, the MTCNN detector was used, with 5 landmark points). For example, the authors of this repository (and maybe authors of generic ArcFace) used following alignment algorithm:

```
src = np.array([
      [30.2946, 51.6963],
      [65.5318, 51.5014],
      [48.0252, 71.7366],
      [33.5493, 92.3655],
      [62.7299, 92.2041] ], dtype=np.float32 )
if image_size[1]==112:
      src[:,0] += 8.0
dst = landmark.astype(np.float32)
tform = trans.SimilarityTransform()
tform.estimate(dst, src)
```
[link](https://github.com/deepinsight/insightface/blob/4a4b8d03fec981912fdef5b3232a37a827cbeed6/src/common/face_preprocess.py#L61-L72) to a source file 
_Notice:_ After this phase your data (cropped faces) will be 112x112 aligned face images. It is convenient to preprocess all the train/val data before the training and validation phase. Since I/O bound will be high, if you want to detect/crop/align and train on the fly.

3. Face embedding extraction phase. This is the core process of face recognition, here the model does a forward pass with input cropped image and outputs the (1x512) face embedding, which corresponds to the person image in a high dimension space (512-d). In order to make a comparison between two images (either between person A and person A, or person A and person B), we need to get their (1x512) embeddings and compare them with particular similarity/distance metric. As a rule of thumb, cosine similarity is chosen as default. Note that, cosine similarity should be applied only on the L-2 normalized embeddings.

I hope, I described the most important aspects of training process. I will update my personal ArcFace implementation repository soon. Stay tuned. [Link to repo](https://github.com/beybars1/ArcFace_PyTorch.git)

**Comment by karmelyoei** [comment]
@beybars1  Thank you so much for your explanation cant wait to see your repo!

**Comment by Zahra7143** [comment]
I'm still unable to train it on my dataset. Does anyone have any tutorials or resources that could help me with this?

**Comment by sohan-addagudi123** [comment]
How do we set up and use our own validation set?
